### PR TITLE
issue: 3623390 Implementing DOCA Flow steering creation

### PR DESCRIPTION
**Pull request by AlexanderGrissik**
## Description
Adds RX steering through DOCA and enables receiving first packets with multi process application.
This PR is based on PR 138.
This PR requires DOCA 2.7, commit <17ad5454c600b487bb5e7b63b5d406f196f48922 ([devtools]: fix error in binary compatibility script)> with the attached patch.
[0001-TEMP-Enable-Multiprocess-DPDK.patch](https://github.com/Mellanox/libxlio/files/15446734/0001-TEMP-Enable-Multiprocess-DPDK.patch)

##### What
1. Adds Steering through DOCA
2. Adds DOCA RXQ creation/destruction and initial usage
3. Enables fork() to work with DOCA/DPDK for Nginx

##### Why ?
DOCA integration

## Change type
What kind of change does this PR introduce?
- [ ] Bugfix
- [X] Feature
- [ ] Code style update
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Tests
- [ ] Other

## Check list
- [X] Code follows the style de facto guidelines of this project
- [X] Comments have been inserted in hard to understand places
- [ ] Documentation has been updated (if necessary)
- [ ] Test has been added (if possible)

### CODE REVIEW

File: src/core/dev/hw_queue_rx.cpp [hunk 1]
```diff
@@ -519,27 +520,24 @@ bool hw_queue_rx::prepare_rq(uint32_t cqn)
...
+        PRINT_DOCA_ERR(hwqrx_logerr, err1, "doca_eth_rxq_cap_get_max_burst_size");
+        PRINT_DOCA_ERR(hwqrx_logerr, err2, "doca_eth_rxq_cap_get_max_packet_size");
         return false;
     }
 
-    if (type_supported == 0) {
-        hwqrx_logerr("DOCA RXQ CPU-path REGULAR-type is not supported");
-        return false;
-    }
+    hwqrx_loginfo("RXQ caps MaxBurstSize %u, MaxPacketSize %u, Dev:%s", max_burst_size,
```
[the remark this answers is not in the document]
Comment (AlexanderGrissik) [reply src/core/dev/hw_queue_rx.cpp:653]: Sure. We will revisit all the info logs one we have a more stable version. Now it is convenient for debugging on initial stages of DOCA integration.

File: src/core/util/sys_vars.cpp [hunk 2]
```diff
@@ -2005,6 +2005,13 @@ void mce_sys_var::get_env_params()
 
 void set_env_params()
 {
+#if defined(DEFINED_NGINX)
+    // We must not call setenv for Nginx master process.
```
[the remark this answers is not in the document]
Comment (AlexanderGrissik) [reply src/core/util/sys_vars.cpp:2009]: Currently it is only for DOCA. But looks like it is needed also for regular XLIO. Today it seems to work by luck in regular XLIO.

File: src/core/util/sys_vars.h [hunk 3]
```diff
@@ -799,7 +799,7 @@ extern mce_sys_var &safe_mce_sys();
 #define MCE_DEFAULT_CQ_AIM_MAX_PERIOD_USEC         (250)
 #define MCE_DEFAULT_CQ_AIM_INTERVAL_MSEC           (250)
 #define MCE_DEFAULT_CQ_AIM_INTERRUPTS_RATE_PER_SEC (5000)
-#define MCE_DEFAULT_CQ_POLL_BATCH                  (16)
+#define MCE_DEFAULT_CQ_POLL_BATCH                  (128)
```
[the remark this answers is not in the document]
Comment (AlexanderGrissik) [reply src/core/util/sys_vars.h:802]: This works well for now. And this is the nginx default anyway. We can do the tuning later.

File: src/core/dev/hw_queue_rx.cpp [hunk 4]
```diff
@@ -551,13 +668,35 @@ bool hw_queue_rx::prepare_rq(uint32_t cqn)
...
+    }
 
     err = doca_eth_rxq_set_type(m_doca_rxq.get(), DOCA_ETH_RXQ_TYPE_REGULAR);
     if (DOCA_IS_ERROR(err)) {
         PRINT_DOCA_ERR(hwqrx_logerr, err, "doca_eth_rxq_get_type_supported");
         return false;
     }
 
+    err = doca_eth_rxq_task_recv_set_conf(m_doca_rxq.get(), task_completion_cb, rx_task_error_cb,
+                                          safe_mce_sys().cq_poll_batch_max);
```
Comment (iftahl) [inline src/core/dev/hw_queue_rx.cpp:687]: why do we use here "cq_poll_batch_max"?
if I understand it correctly - we should be able to post big number of tasks to fill the rq, but cq_poll_batch_max refers to how many wqes to poll we do for each polling.

File: src/core/dev/rfs_rule.cpp [hunk 5]
```diff
@@ -88,6 +99,41 @@ bool rfs_rule::create(dpcp::match_params &match_value, dpcp::match_params &match
                *reinterpret_cast<uint64_t *>(match_value.dst.ipv6),
                *reinterpret_cast<uint64_t *>(match_value.dst_mac));
 
+    if (0) { // TEMP: Switch to 1 to enable DOCA Flow
```
Comment (iftahl) [inline src/core/dev/rfs_rule.cpp:102]: you had a param for that before, don't you want it here?

Comment (AlexanderGrissik) [reply src/core/dev/rfs_rule.cpp:102]: This param will be added in RX path PR. Here it's just to be able to turn on for debugging. RX data path is not working yet. Only receiving a few packets.

File: src/core/dev/hw_queue_rx.cpp (hunk 4 above)
Comment (AlexanderGrissik) [reply src/core/dev/hw_queue_rx.cpp:687]: This is just temporary. It will be changed.

**Review by iftahl — APPROVED** [review]

**Pull request merged**